ios/browser/api/brave_wallet/brave_wallet_api.mm

```diff
@@ -0,0 +1,185 @@
+#endif
+
+@implementation BraveWalletAPI {
+  ChromeBrowserState* _mainBrowserState;  // NOT OWNED
+  NSMutableDictionary<NSNumber* /* BraveWalletCoinType */,
+                      NSDictionary<BraveWalletProviderScriptKey, NSString*>*>*
+      _providerScripts;
+}
+
+- (instancetype)initWithBrowserState:(ChromeBrowserState*)mainBrowserState {
+  if ((self = [super init])) {
+    _mainBrowserState = mainBrowserState;
+    _providerScripts = [[NSMutableDictionary alloc] init];
+  }
+  return self;
+}
+
++ (id<BraveWalletBlockchainRegistry>)blockchainRegistry {
+  auto* registry = brave_wallet::BlockchainRegistry::GetInstance();
+  return [[BraveWalletBlockchainRegistryImpl alloc]
+      initWithBlockchainRegistry:registry];
+}
+
+- (nullable id<BraveWalletEthereumProvider>)
+    ethereumProviderWithDelegate:(id<BraveWalletProviderDelegate>)delegate
+               isPrivateBrowsing:(bool)isPrivateBrowsing {
+  auto* browserState = _mainBrowserState;
+  if (isPrivateBrowsing) {
+    browserState = browserState->GetOffTheRecordChromeBrowserState();
+  }
+
+  auto* json_rpc_service =
+      brave_wallet::JsonRpcServiceFactory::GetServiceForState(browserState);
+  if (!json_rpc_service) {
+    return nil;
+  }
+
+  auto* tx_service =
+      brave_wallet::TxServiceFactory::GetServiceForState(browserState);
+  if (!tx_service) {
+    return nil;
+  }
+
+  auto* keyring_service =
+      brave_wallet::KeyringServiceFactory::GetServiceForState(browserState);
+  if (!keyring_service) {
+    return nil;
+  }
+
+  auto* brave_wallet_service =
+      brave_wallet::BraveWalletServiceFactory::GetServiceForState(browserState);
+  if (!brave_wallet_service) {
+    return nil;
+  }
+
+  auto* provider = new brave_wallet::EthereumProviderImpl(
+      ios::HostContentSettingsMapFactory::GetForBrowserState(browserState),
+      json_rpc_service, tx_service, keyring_service, brave_wallet_service,
+      std::make_unique<brave_wallet::BraveWalletProviderDelegateBridge>(
+          delegate),
+      browserState->GetPrefs());
+  return [[BraveWalletEthereumProviderImpl alloc]
+      initWithEthereumProvider:provider];
+}
+
+- (nullable id<BraveWalletSolanaProvider>)
+    solanaProviderWithDelegate:(id<BraveWalletProviderDelegate>)delegate
+             isPrivateBrowsing:(bool)isPrivateBrowsing {
+  auto* browserState = _mainBrowserState;
+  if (isPrivateBrowsing) {
+    browserState = browserState->GetOffTheRecordChromeBrowserState();
+  }
+
+  auto* keyring_service =
+      brave_wallet::KeyringServiceFactory::GetServiceForState(browserState);
+  if (!keyring_service) {
+    return nil;
+  }
+
+  auto* brave_wallet_service =
+      brave_wallet::BraveWalletServiceFactory::GetServiceForState(browserState);
+  if (!brave_wallet_service) {
+    return nil;
+  }
+
+  auto* tx_service =
+      brave_wallet::TxServiceFactory::GetServiceForState(browserState);
+  if (!tx_service) {
+    return nil;
+  }
+
+  auto* provider = new brave_wallet::SolanaProviderImpl(
```

Comment:
This leaks.
The `brave_wallet::SolanaProviderImpl` takes a pointer, stores it as a raw_ptr but it never deletes it for you

Reply:
Fixed in https://github.com/brave/brave-core/pull/14703/commits/880af8f9b696d95a7013ac4e8e37ced6089b1e4c